### ISSUE TITLE
V93k testmethodlimits section in .tf file is incorrect for SpecSearch

### ISSUE THREAD

**Issue by jrwilson77** [Origen-SDK/origen#251]
Hi,

After loading my test program I'm getting a 93k SmarTest inconsistency error for a SpecSearch Test Method. Origen inserts a parameter labeled Functional in the testmethodlimits section of a .tf file when it should be SpecSearch_Test. I add the correct parameter to my .tf file the inconsistency is resolved.

Thanks, 
Jim

SmarTest Inconsistency Check window with incorrect parameter identified.
![origen_bug](https://user-images.githubusercontent.com/37595127/37677272-9272739c-2c48-11e8-989e-c0742425a778.JPG)

**Comment by ginty** [comment]
Hi @jrwilson77, thanks for reporting this.

I had a look at how it works and if no limit name is defined then it falls back to 'Functional', this follows the convention laid down by the Advantest AC and DC TMLs.

To change it, the `test_name` attribute should be used to define an alternative default in your [Custom Test Method Definition](http://origen-sdk.org/origen/guides/program/v93k/#Custom_Test_Methods).
e.g.

~~~ruby
spec_search: {
  test_name: 'SpecSearch_Test',
  #...
}
~~~

Alternatively, you can also override it on a per test instance basis within your test program interface code like this:

~~~ruby
tm = test_methods.my_tml.spec_search
tm.test_name = 'SpecSearch_Test'
~~~

In summary, I don't think this is a bug in Origen but rather in your application's definition of your TML (which is missing the test_name attribute), but perhaps we should take the action to make the documentation clearer in this respect.

cc @info-rchitect

**Comment by info-rchitect** [comment]
@ginty @jrwilson77  I will check out TML importer and also take the action to update the docs after I get a fix.

**Comment by info-rchitect** [comment]
OK the issue is that the custom test methods do not set the 'testName' parameter as expected.  Here is an Advantest SpecSearch method:

~~~text
testmethod 30 -- ac_tml.AcTest.SpecSearch
  location name: "SpecSearch.cpp:SpecSearch"
  shared lib path: "/opt/hp93000/soc/segments/tml/sh_lib/libac_tml.so"
  source project path: ""
  compiled with SmarTest version: "7.3.2.11, 05-Jul-16"
  parameter ids:
    result0 :double, 0.0
    result1 :double, 0.0
    result2 :double, 0.0
    result3 :double, 0.0
    spec :SpecVariable, 
    setupPinlist :PinString, 
    method :string, linear
    min :double, 9999
    max :double, 9999
    step :string, 
    resolution :string, 
    resultPinlist :PinString, 
    output :string, None
    testName :string, SpecSearch_Test
  limit ids:
    SpecSearch_Test :NA, 0, NA, 0, 
~~~

However, there is a workaround by grabbing the limit_ids information:

~~~text
  limit ids:
    SpecSearch_Test :NA, 0, NA, 0, 
~~~

@jrwilson77 I will send you some code to check out a fix.

**Comment by info-rchitect** [comment]
@ginty is the test_name attribute at the same level as the class_name attribute or is it supposed to be included in the test_method parameter hash?

**Comment by info-rchitect** [comment]
@jrwilson77 I sent you an email with a commit tag to test out the fixed test interface.  If it works, please say so here and close the issue.

**Comment by jrwilson77** [comment]
Didn't fix the issue.  Moved it around a bit though.  There's still an inconsistency check error but this time not because the testmethodlimits entry for a SpecSearch test is incorrect, that looks fine now but rather because there is an extra test parameter: "testName" = "SpecSearch_Test" that is not in our test method library.  If I remove this entry manually and re-load the test program the inconsistency error goes away.

**Comment by info-rchitect** [comment]
@ginty I think this poses a problem because it seems that origen_testers needs the test_name parameter to set the limits name correctly.  Are there any other options for our test interface to fix this?

**Comment by ginty** [comment]
Looking at the screenshot from @jrwilson77, is the problem that the TML has called this 'TestName' instead of 'testName'?

Also to confirm, we are talking about a custom SpecSearch test method here and not `ac_tml.AcTest.SpecSearch`, right?

I believe that if you prefix the parameter name with an underscore in the test method definition then it will convert to the camelcase format with the first character upper cased:

~~~ruby
spec_search: {
  _test_name: [:string, 'SpecSearch_Test']
}
~~~

alternatively, you could also try a string:

~~~ruby
spec_search: {
  'TestName' => [:string, 'SpecSearch_Test']
}
~~~

However, you may need to combine that with an update to this Origen method to make it look for alternative formatting of the `test_name` attribute: https://github.com/Origen-SDK/origen_testers/blob/master/lib/origen_testers/smartest_based_tester/base/test_methods/limits.rb#L57

e.g. 

~~~ruby
name = test_method.try(:test_name) || test_method.try(:_test_name) || test_method.try('TestName')
~~~

**Comment by info-rchitect** [comment]
@ginty this is a custom SpecSearch test method.  I will try your suggestions.

**Comment by info-rchitect** [comment]
@ginty just to be clear, this should work if our test method has the :test_name set to 'SpecSearch_Test'?

**Comment by ginty** [comment]
Not sure what you mean by that given all the previous discussion in this thread...but yes, apart from the fact that as a side effect it generates a test method attribute called 'testName' which your app doesn't like.

**Comment by info-rchitect** [comment]
@ginty I implemented the code you suggested and tried all of the combinations.  Each one worked to fix the limits name but all of them left a test method parameter ('testName', 'TestName') that the test method cannot use.  If I understand the issue correctly, Origen testers currently forces test methods to include a 'TestName' parameter if they want to set the limits name to anything other than 'Functional'.  I don't think this is a limitation in our TML, rather a design issue with Origen testers.  Some TML (e.g. ac_tml) put a 'testName' parameter in their test methods so the limits get set correctly.

~~~text
testmethod 30 -- ac_tml.AcTest.SpecSearch
  location name: "SpecSearch.cpp:SpecSearch"
  shared lib path: "/opt/hp93000/soc/segments/tml/sh_lib/libac_tml.so"
  source project path: ""
  compiled with SmarTest version: "7.3.2.11, 05-Jul-16"
  parameter ids:
    result0 :double, 0.0
    result1 :double, 0.0
    result2 :double, 0.0
    result3 :double, 0.0
    spec :SpecVariable, 
    setupPinlist :PinString, 
    method :string, linear
    min :double, 9999
    max :double, 9999
    step :string, 
    resolution :string, 
    resultPinlist :PinString, 
    output :string, None
    testName :string, SpecSearch_Test
  limit ids:
    SpecSearch_Test :NA, 0, NA, 0, 
~~~

But others, including ours, do not have test name as a parameter.

~~~text
testmethod 20 -- multi_group_dc_tml.DcTest.Leakage
  location name: "Leakage.cpp:Leakage"
  shared lib path: "/opt/hp93000/soc/segments/tml/sh_lib/libmulti_group_dc_tml.so"
  source project path: ""
  compiled with SmarTest version: "7.3.2.11, 05-Jul-16"
  parameter ids:
    UsePPMU :string, YES
    UsePPMU.pinlist :PinString, 
    UsePPMU.measure :string, BOTH
    UsePPMU.forceVoltage_mV :string, (400,3800)
    UsePPMU.preCharge :string, ON
    UsePPMU.prechargeVoltage_mV :string, (0.0, 0.0)
    UsePPMU.settlingTime_ms :string, (1.0,1.2)
    UsePPMU.measureMode :string, PAR
    UsePPMU.relaySwitchMode :string, DEFAULT(BBM)
    UsePPMU.testName :string, 
    UseSPMU :string, NO
    UseSPMU.pinlist :PinString, 
    UseSPMU.measure :string, BOTH
    UseSPMU.forceVoltage_mV :string, (400,3800)
    UseSPMU.preCharge :string, ON
    UseSPMU.prechargeVoltage_mV :string, (0.0, 0.0)
    UseSPMU.clampCurrent_uA :string, (100,300)
    UseSPMU.settlingTime_ms :string, (1.0,1.2)
    UseSPMU.measureMode :string, PAR
    UseSPMU.relaySwitchMode :string, NTBBM
    UseSPMU.testName :string, 
    UseMCX-PMU :string, NO
    UseMCX-PMU.pinlist :PinString, 
    UseMCX-PMU.measure :string, BOTH
    UseMCX-PMU.forceVoltage_mV :string, (400,3800)
    UseMCX-PMU.preCharge :string, ON
    UseMCX-PMU.prechargeVoltage_mV :string, (0.0, 0.0)
    UseMCX-PMU.measureMode :string, PAR
    UseMCX-PMU.settlingTime_ms :string, (1.0,1.2)
    UseMCX-PMU.testName :string, 
    UseDcScaleSIG :string, NO
    UseDcScaleSIG.pinlist :PinString, 
    UseDcScaleSIG.measure :string, BOTH
    UseDcScaleSIG.forceVoltage_mV :string, (400,3800)
    UseDcScaleSIG.preCharge :string, ON
    UseDcScaleSIG.prechargeVoltage_mV :string, (0.0,0.0)
    UseDcScaleSIG.clampCurrent_uA :string, (100,300)
    UseDcScaleSIG.settlingTime_ms :string, (1.0,1.2)
    UseDcScaleSIG.measureMode :string, PAR
    UseDcScaleSIG.relaySwitchMode :string, NO
    UseDcScaleSIG.testName :string, 
    preFunction :string, NO
    preFunction.testName :string, 
    stopVecLow :int, 
    stopVecHigh :int, 
    stopCycLow :string, 
    stopCycHigh :string, 
    checkFunction :string, ON
    output :string, None
  limit ids:
    PreFunctionalTest :NA, 0, NA, 0, 
    PreFunctionalTest_Low :NA, 0, NA, 0, 
    PreFunctionalTest_High :NA, 0, NA, 0, 
  testnumber ids:
~~~

Shouldn't there be a way for the limits name to be set, regardless of what parameters the test method contains?

**Comment by ginty** [comment]
_Shouldn't there be a way for the limits name to be set, regardless of what parameters the test method contains?_

Seems like there needs to be.

**Comment by info-rchitect** [comment]
@ginty OK I will take a shot.  I am thinking to allow options[:limits_id] to be passed as a test method parameter and it would set the limits name, taking precedence over the test name parameter.  If passed, it would be deleted before passing the test method parameters to the generator.

**Comment by info-rchitect** [comment]
[here is a fix](https://github.com/Origen-SDK/origen_testers/pull/85)

**Comment by info-rchitect** [comment]
[closed](https://github.com/Origen-SDK/origen_testers/pull/85)
